File: src/egypt_research_mcp/exports.py
from __future__ import annotations

import csv
import io
import json
import re

from .models import SearchResult


FORMATS = {"csv", "jsonl", "bibtex", "ris"}


def _year(result: SearchResult) -> str:
    return str(result.published_at.year) if result.published_at else ""


def export_results(results: list[SearchResult], format_name: str) -> str:
    if format_name not in FORMATS:
        raise ValueError(f"Unsupported export format: {format_name}")
    if format_name == "jsonl":
        return "\n".join(
            json.dumps(result.model_dump(mode="json"), ensure_ascii=False)
            for result in results
        )
    if format_name == "csv":
        output = io.StringIO()
        writer = csv.writer(output)
        writer.writerow(
            ["title", "source", "source_type", "published_at", "url", "topics"]
        )
        for result in results:
            writer.writerow(
                [
                    result.title,
                    result.source_name,
                    result.source_type,
                    result.published_at.isoformat() if result.published_at else "",
                    result.canonical_url,
                    "; ".join(result.topics),
                ]
            )
        return output.getvalue()
    if format_name == "bibtex":
        records = []
        for result in results:
            key = re.sub(r"[^a-zA-Z0-9]", "", result.source_slug) or "egypt"
            records.append(
                "\n".join(
                    [
                        f"@misc{{{key}{result.document_id},",
                        f"  title = {{{result.title}}},",
                        f"  author = {{{result.source_name}}},",
                        f"  year = {{{_year(result)}}},",
                        f"  url = {{{result.canonical_url}}}",
                        "}",
                    ]
                )
            )
        return "\n\n".join(records)
    records = []
    for result in results:
        lines = [
            "TY  - ELEC",
            f"TI  - {result.title}",
            f"AU  - {result.source_name}",
            f"PY  - {_year(result)}",
            f"UR  - {result.canonical_url}",
            "ER  -",
        ]
        records.append("\n".join(lines))
    return "\n\n".join(records)

File: src/exports.ts
import type { SearchResult } from "./types.js";

export const EXPORT_FORMATS = ["csv", "jsonl", "bibtex", "ris"] as const;
export type ExportFormat = (typeof EXPORT_FORMATS)[number];

const csv = (value: unknown): string => `"${String(value ?? "").replaceAll('"', '""')}"`;
const key = (result: SearchResult): string => `egypt${result.documentId}`;

export function exportResults(results: SearchResult[], format: ExportFormat): string {
  if (format === "jsonl") return results.map((result) => JSON.stringify(result)).join("\n");
  if (format === "csv") {
    const header = "document_id,title,source_name,source_type,url,published_at,archived_at";
    return [header, ...results.map((result) => [result.documentId, result.title, result.sourceName, result.sourceType, result.canonicalUrl, result.publishedAt ?? "", result.archivedAt].map(csv).join(","))].join("\n");
  }
  if (format === "bibtex") return results.map((result) => `@misc{${key(result)},\n  title = {${result.title}},\n  author = {${result.sourceName}},\n  url = {${result.canonicalUrl}},\n  urldate = {${result.archivedAt.slice(0, 10)}}\n}`).join("\n\n");
  return results.map((result) => ["TY  - ELEC", `TI  - ${result.title}`, `AU  - ${result.sourceName}`, `UR  - ${result.canonicalUrl}`, result.publishedAt ? `PY  - ${result.publishedAt.slice(0, 10)}` : null, `Y2  - ${result.archivedAt.slice(0, 10)}`, "ER  -"].filter(Boolean).join("\n")).join("\n\n");
}
